File: typescript/src/com/murex/tbw/domain/book/Novel.ts
import { Language } from "../country/Language";
import { Author } from "./Author";
import { Book } from "./Book";
import { Genre } from "./Genre";
import * as Immutable from "immutable";
import equal = require("fast-deep-equal");
import hash from "hash-it";

export function isNovel(book: Book): boolean {
  return (book as Novel).genres !== undefined;
}

export class Novel implements Book {
  readonly name: string;
  readonly price: number;
  readonly author: Author;
  readonly language: Language;
  readonly genres: Immutable.Set<Genre>;

  public constructor(
    name: string,
    price: number,
    author: Author,
    language: Language,
    genres: Immutable.Set<Genre>
  ) {
    this.name = name;
    this.price = price;
    this.author = author;
    this.language = language;
    this.genres = genres;
  }

  equals(other: any): boolean {
    return equal(this, other);
  }
  hashCode(): number {
    return hash(this);
  }
}

File: kotlin/src/main/kotlin/com/murex/tbw/domain/book/Novel.kt
/*******************************************************************************
 *
 * Copyright (c) {2003-2021} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the MIT License
 * which accompanies this distribution, and is available at
 * https://opensource.org/licenses/MIT
 *
 *******************************************************************************/

package com.murex.tbw.domain.book

import com.murex.tbw.domain.country.Language

data class Novel(
    override val name: String,
    override val price: Double,
    override val author: Author,
    override val language: Language,
    val genres: List<Genre> = mutableListOf()

) : Book